[2단계 - 상품 목록] 니야(최수연) 미션 제출합니다.

sooyeoniya opened the pull request:
안녕하세요, 해먼드! 니야입니다 😁 날씨 좋은 월요일이네요 ☀

1단계에서 추가로 달아주신 PR 내용도 다 읽어봤습니다!
거기서 반영된 사항들도 있고, 추가로 궁금한 건 아래 내용에 적어놨습니다 ㅎㅎ 
확인해주시면 감사하겠습니다! 이번 2단계도 잘 부탁드려요 🙇🏻‍♀️

이번에는 MSW를 사용해 응답을 모킹했는데요, 그 과정에서 지난번 서버의 HTTPS 이슈도 함께 해결해놔서, 배포 페이지 접근에는 문제가 없을 거예요. 다만 로컬에서 테스트하신다면, `VITE_USE_MOCK=true` 설정을 추가해주셔야 해서 이 부분 참고 부탁드립니다!

```
// 위에 두 개는 지난번 슬랙으로 부분들입니다.
VITE_API_BASE_URL=http://techcourse-lv2-alb-974870821.ap-northeast-2.elb.amazonaws.com/
VITE_TOKEN=XX
VITE_USE_MOCK=true // 이번 2단계에서 추가된 부분입니다.
```

# 📦 상품 목록

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

## 1단계

- 비동기 작업(API 요청)의 상태를 관리하고, 비동기 요청 상태에 따라 적절한 UI를 보여줄 수 있다.
- React Testing Library(RTL)를 활용하여 주요 기능에 대한 테스트를 작성할 수 있다.

## 2단계

- Single Source Of Truth 를 지향하며 일관된 인터페이스를 가진 Data fetching hook 을 구현한다.
- MSW를 이용해 서버의 다양한 응답을 시뮬레이션하고, 그 시나리오에 따른 UI 처리 및 테스트를 구현할 수 있다.

# 🕵️ 셀프 리뷰(Self-Review)

## 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] RTL 테스트 케이스를 모두 작성했나요?
- [x] API 요청 시 로딩/예외 케이스를 전부 고려했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포 링크 기입: [니야의 SHOP 🛍️](https://sooyeoniya.github.io/react-shopping-products/)

## 리뷰 요청 & 논의하고 싶은 내용

## 1) Data fetching hook 설계 의도

범용적인 Context API를 만들어, 하나의 Provider로 모든 API 데이터를 관리하는 훅을 구현해봤습니다.

<img width="843" alt="image" src="https://github.com/user-attachments/assets/15b5266b-5fff-47af-bdc6-7398977f2704" />


## 2) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

### ✅ MSW와 함께 배포 시 HTTPS 에러 해결 방법

앞서 말씀드렸듯이, 지난 1단계에서는 서버가 HTTP였기 때문에 클라이언트가 사용하는 HTTPS와 프로토콜이 달라 네트워크 관련 오류가 발생했었습니다. 이로 인해 브라우저에서 ‘안전하지 않은 사이트’를 허용하는 설정이 필요했는데요.

이번에는 크루인 `피터`가 슬랙에 공유해준 정보를 참고하여 이 문제를 한 번 해결해보았습니다!
MSW를 사용하고 있기 때문에, 서버 URL을 임의의 HTTPS로 시작하는 주소로 변경해주는 방식으로 해결할 수 있었습니다.

물론 실제 서버를 사용하는 경우에는 문제가 될 수 있지만, 이번 프로젝트에서는 프로덕션 환경에서도 MSW 서버를 사용하도록 설정했기 때문에 별다른 문제 없이 정상적으로 작동합니다!

### ✅ 모달 설계 개선

모달을 최대한 재사용 가능한 구조로 만들고 싶었습니다.
이를 위해 Context API를 사용해 모달을 구현하고, 내부 콘텐츠만 동적으로 바뀔 수 있도록 설계했습니다.

이때, 모달 내부 콘텐츠를 어떻게 바꿀지에 대해 두 가지 방식을 고민했습니다.
1. 모달을 열 때 type을 전달하고, 내부에서 해당 타입에 따라 콘텐츠를 분기 처리하는 방식
2. 모달을 열 때 렌더링할 콘텐츠 자체를 인자로 전달하는 방식

처음에는 1번 방식을 적용해, 모달을 열 때 modalType을 인자로 넘기고, 해당 타입에 맞는 콘텐츠가 렌더링되도록 했습니다.
이때 모달의 열림/닫힘 상태는 modalType의 유무로 판단하며, 모달을 닫을 때는 이를 null로 설정했습니다.

하지만 이 방식에는 몇 가지 문제가 있었는데요,
- CartModal 등 모달 컴포넌트에 전달되는 컨텐츠의 props가 일관되게 관리되지 않아, props 타입이 unknown으로 처리되는 문제가 있었습니다. 이 때문에 모달 type 별로 필요한 props 타입을 정의한 인터페이스가 필요했습니다.
- 더 큰 문제는 OCP(Open–Closed Principle)에 어긋난다는 점이었습니다. 새로운 모달이 추가될 때마다 컨텐츠 분기 처리해주는 ModalProvider의 내부 로직을 수정해야 했기 때문에, 확장에는 열려 있지만, 수정에도 열려 있는 구조가 되었습니다.

따라서, 더 유연한 구조를 위해 **외부에서 모달 콘텐츠를 직접 주입하는 방식(2번 방식)** 으로 변경했고, 이를 통해 OCP를 지키면서도 props 관리 또한 훨씬 깔끔해졌습니다.

### ✅ Products 필터링 상태 관리

범용적인 API Context를 만들면서 데이터를 하나의 상태로 관리하고 있는데, 여기서 products와 관련된 필터링 상태를 어떻게 관리할지 고민이 되었습니다.

API Context 내부에서 필터링 상태를 같이 관리하면, 범용 컨텍스트임에도 잦은 수정이 발생해서 뭔가 역할이 모호해질 것 같았습니다.
또한, 아래와 같이 API Provider에 필터링 상태를 추가하는 것은 앞서 말한 것처럼 범용 컨텍스트의 역할과 맞지 않는 느낌입니다.

```tsx
export const APIProvider = ({ children }: PropsWithChildren) => {
  const [data, setData] = useState<ApiDataState>({});
  const [filters, setFilters] = useState<ApiFilters>(defaultFilters);
```

하지만 반대로, 이 useAPI를 호출해서 데이터를 가공하는 useProducts 훅 내부에서 상태를 따로 만들면 이 필터링 상태가 전역 상태가 되지 못해, useProducts를 사용하는 곳마다 별도의 상태가 생성되어 SSOT(Single Source of Truth) 원칙에 어긋납니다.

따라서, 이 문제를 해결하고자 두 가지 방식을 고민해봤습니다.

1. **data 내부의 products에 필터링 상태를 포함시키기**
    
    ```tsx
    // 상품 데이터 내부에 필터 상태 넣는 구조
    export interface ApiDataTypeMap {
      products: {
        products: Products;     // 상품 데이터
        filters: {              // 필터 상태
          category: CategoryOptionsKey;
          sortOptions: SortOptionsKey;
        };
      };
      cartItems: CartItems;
    }
    ```
    
2. **URL 기반 접근법**
    
    useSearchParams를 활용해 카테고리, 정렬 필터링 정보를 URL 파라미터에 저장하는 방법도 생각했습니다. 사실 이 방식은 Context API를 사용하지 않고도 전역 상태와 유사한 상태 유지 효과를 낼 수 있습니다. 따라서, 이 방식을 채택해서 필터 상태를 관리하도록 했습니다.

## 3) 이번 리뷰를 통해 논의하고 싶은 부분

### 💬 폴더 구조 관련 고민

이번에 구현을 하다보니 폴더 구조가 너무 지저분해서, 이를 정비하면서 지난 영화 미션 리뷰어분께서 추천해주신 **FSD 아키텍처**를 도입해보았는데요, 적용 과정에서 몇 가지 애매한 지점이 있어 해면드의 의견을 듣고 싶습니다!

**1. features를 어떻게 활용해야할지 모르겠습니다.**

컴포넌트와 도메인 관리가 제대로 되지 않아 FSD 구조를 도입하게 되었는데, 적용하면서 features 폴더의 사용이 다소 애매하게 느껴졌습니다. 현재 프로젝트는 대부분 UI 중심이라 기능 단위로 features를 나누기보다는, 오히려 widgets에서 컴포넌트를 계층적으로 정리하는 방식이 더 자연스럽게 느껴졌습니다. 또한, 상품이나 장바구니 아이템과 같은 도메인은 entities에 위치하고, 관련 UI 컴포넌트는 widgets에 depth를 두어 구조화했기 때문에, features의 역할이 더욱 모호하게 느껴졌습니다. 이 구조에서 features를 어떻게 활용해야 할지 잘 감이 오지 않습니다.

**2. Modal / Toast / API 관련 코드 구조가 너무 분산되어 있는 것 같습니다.**

현재 Modal과 Toast 관련 로직은 다음과 같이 여러 폴더에 나뉘어져 있습니다.

- Context: `app/providers`
- 관련 hook: `shared/hooks`
- 관련 상수: `shared/config`
- UI 컴포넌트: `shared/ui`

API 관련 코드도 비슷하게 구조가 분산되어 있습니다.

- Context: `app/providers`
- 관련 hook (`useAPI`): `shared/hooks`
- API 요청 함수들: `shared/api`

이처럼 **하나의 기능에 필요한 코드들이 너무 물리적으로 떨어져 있어**, 오히려 기능 단위로 한 곳에 묶고 싶다는 생각이 들었습니다. 예를 들어, modal, toast처럼 **UI + 상태 + 로직이 섞인 공통 기능**은 features 디렉토리로 모아 관리하는 방식이 더 직관적이지 않을까 고민 중입니다.

하지만 여기서 고민인 것은, FSD 구조를 보면, 

- `app`과 `shared`는 **역할 단위**
- `features`는 **기능 단위**

이런 느낌이 강하게 드는데요. 그래서 더 폴더 구분 기준이 혼재되어 있다는 느낌도 듭니다. 
이런 경우 보통 어떻게 구조를 잡는 것이 좋을지, 해먼드의 의견이 궁금합니다.

### 💬 ‘is’ 접두사 사용에 대한 고민

지난 번 리뷰에서 말씀하신대로 변수명이나 함수명에서 ‘is’ 접두사를 모두 제거하려고 했습니다. 그런데 아래 예시처럼 ‘isValidUrl’ 같은 경우는 ‘유효한 URL인가?’라는 의미를 가장 잘 나타내는 이름이라고 생각합니다. ‘checkValidUrl’보다 훨씬 적합해 보인다고 생각했어요.

```tsx
const isValidUrl = (url: string): boolean =>
  Boolean(url && (url.startsWith("http://") || url.startsWith("https://")));

export const getImageUrl = (url: string): string =>
  isValidUrl(url) ? url : DEFAULT_IMAGE_URL;
```

여기서 한 가지 궁금한 점이 있었는데요. ‘is’ 접두사를 쓰지 않는 이유가 함수와 변수를 구분하기 어려워서 그런 걸까요? 저는 위와 같은 케이스의 경우, 함수명에는 ‘is’를 붙여도 괜찮다고 생각합니다. 해먼드님께서는 ‘is’ 접두사를 아예 사용하지 않으시는 것 같은데, 혹시 이유가 정확히 무엇인지 궁금합니다!

### 💬 테스트에서 API 모킹과 실제 호출의 딜레마

테스트를 구현하면서 고민이 있었습니다. 저번에 말씀해주신 대로, 모킹한 로직 자체를 테스트하는 것은 큰 의미가 없다는 점에 저도 동의합니다. 그런데 이번에도 두 개의 훅을 테스트하는 과정에서 결국 모킹을 하게 되더라고요 🥲

특히 MSW 서버(msw/node)를 사용해 내부적으로 테스트를 돌렸는데, 함수 호출 후에 refetch가 제대로 실행됐는지 `toHaveBeenCalled()`로 확인하려 해도 오류가 발생했습니다. 실제로 상품 추가, 삭제, 수량 업데이트 함수들을 호출해도 데이터가 갱신되지 않았습니다.

결과적으로, MSW 응답 자체를 테스트 내에서 모킹해 진행하고 있습니다. 하지만, 필터링과 정렬이 포함된 `useProducts` 훅은 실제 API를 호출해 데이터가 잘 업데이트되는지도 별도로 확인하는 테스트를 추가했습니다.

제가 고민하는 부분은, `useCartItems` 같은 훅에서 다양한 함수들을 사용하는데, API 응답을 모킹하지 않고 제대로 테스트하는 방법을 찾기 어렵다는 점입니다. 제가 잘 활용을 못하고 있는 것 같아서 어떻게 하면 응답 모킹 없이 테스트 오류가 나지 않고 데이터 갱신이 잘 되는 것을 확인할 수 있는지 궁금합니다!

아니면 아예 이런 경우에는 API를 직접 호출하는 통합 테스트를 진행하고, 별도로 UI가 잘 업데이트되는지 화면 기반 테스트를 분리해서 하는 것이 더 나을까요?

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. API 요청

- Data Fetching hook 의 필요성을 느끼고, 본인만의 설계 의도를 가지고 있는가?
- Data Fetching hook 이 확장성 있는 인터페이스를 가지는가?
- SSOT 를 지키는가?

### 2. MSW/Test

- 서버 API 가 나오지 않았지만, 요구사항으로 주어진 기능을 MSW 를 활용하여 구현하였는가?

sooyeoniya committed:
docs: 2단계 요구사항 README 문서 정리

sooyeoniya committed:
feat: 수량 증감 버튼 UI 구현

sooyeoniya committed:
design: 이미지, 아이콘 요소 드래그 방지 추가

sooyeoniya committed:
chore: 사용하지 않는 SelectBox id 제거

sooyeoniya committed:
chore: ButtonIcon, CartIcon에 접근성 향상을 위한 속성 수정

sooyeoniya committed:
feat: 사용하지 않는 빼기 버튼 제거

sooyeoniya committed:
feat: 담고 빼기 토글 기능 분리

- 증가, 감소 기능 분리하여 버튼에 별개로 작동하도록 수정

sooyeoniya committed:
refactor: 상품 목록 스켈레톤 분리

sooyeoniya committed:
feat: 아이템 수량 증가 감소 로직 API 연동

patch API 추가 및 각각의 클릭 기능 연동

sooyeoniya committed:
feat: ModalContext 구현

[세부사항]
- 모달위치의 부모 요소에 custom-root라는 id를 넣고, createPortal을 사용해 해당 위치에 모달을 걸어줌
- 모달 껍데기를 재사용할 수 있도록 하고, 모달 내부 컨텐츠만 바뀔 수 있도록 구현함
- ShopHeader의 장바구니 아이콘과 상품 개수 표시하는 부분을 하나의 CartButton으로 묶어서 관리함
- CartButton을 클릭했을 때 모달이 열리도록 함
- 모달 열림/닫힘 여부는 modalType을 사용해 관리하도록 함

sooyeoniya committed:
design: 상품 아이템 버튼 위치 CSS 속성 상위 요소로 분리

스타일 독립성 보장

sooyeoniya committed:
design: 장바구니 내부 컨텐츠 UI 구현

sooyeoniya committed:
feat: 장바구니 목록 데이터 연동

[세부 사항]
모달 컨텍스트 내부에서 내부 내용을 분기처리하는 것이 아닌, 내부 내용을 외부에서 주입받아서 관리하도록 수정

sooyeoniya committed:
chore: 금액 단위 컨벤션 맞춰 표시

sooyeoniya committed:
feat: 장바구니에 담긴 목록들만 필터링

sooyeoniya committed:
design: 이미지 경로가 잘못된 경우 default image 적용

sooyeoniya committed:
docs: useProducts TODO 작성

sooyeoniya committed:
feat: Toast Context API 로 변경

에러 메시지만 넘길 수 있도록 수정

sooyeoniya committed:
feat: Toast 메시지 종류에 따라 색상 타입 지정

sooyeoniya committed:
feat: MSW 서버 설정 및 응답 모킹

sooyeoniya committed:
refactor: 장바구니 상품목록들 products가 아닌 cartItems로 변경

sooyeoniya committed:
test: MSW 테스트를 위한 설정

sooyeoniya committed:
chore: cartItemCount -> cartItemsCount로 변경

sooyeoniya committed:
feat: quantityByCartId 함수 제거

sooyeoniya committed:
chore: 상품 이미지 삽입

sooyeoniya committed:
feat: 장바구니에서 수량감소 시 이미 1개인 경우 메시지 출력

sooyeoniya committed:
design: 토스트 타입별 색상 변경

sooyeoniya committed:
feat: 장바구니에서 상품 개수가 최소, 최대에 도달하는 경우 버튼 비활성화

sooyeoniya committed:
fix: 수량 감소 시 1개만 남은 경우 patch가 아닌 delete API 호출

sooyeoniya committed:
feat: 상품 최대 수량 기능 추가

sooyeoniya committed:
feat: 상품 개수가 0개인 경우 품절 표시

sooyeoniya committed:
fix: 장바구니에서 수량 감소 가능하도록 수정

- 원래는 disabled 처리해놨는데, 그냥 1개일 때 수량 감소하면 그냥 장바구니에서 없어지도록 함

sooyeoniya committed:
fix: base url 설정하여 shop 아이콘 클릭시 base 경로로 이동되도록 수정

sooyeoniya committed:
fix: MSW statusText 관련 오류 수정

MSW 오류 해결: String contains non ISO-8859-1 code point
이 오류는 MSW에서 HTTP 응답 메시지에 ISO-8859-1 문자 집합에 포함되지 않은 문자가 포함되어 있을 때 발생

statusText가 한글이어서 발생한 오류인 것으로 추정
-> statusText를 영문으로 수정

sooyeoniya committed:
chore: console.log 제거

sooyeoniya committed:
fix: 상품 기본 이미지 경로 수정

절대경로 -> 상대경로

sooyeoniya committed:
refactor: FSD 아키텍처 기반으로 전체 폴더 구조 변경

sooyeoniya committed:
refactor: useProducts, useCartItems 훅 Context API로 수정

sooyeoniya committed:
entities/model 내부 구조 개편

providers와 types 폴더로 구분

sooyeoniya committed:
refactor: 이미지 처리 로직 유틸로 분리

sooyeoniya committed:
refactor: BASE_URL 상수 가져다 사용

sooyeoniya committed:
refactor: deleteProductInCart 인자로 productId가 아닌 cartId로 대채

sooyeoniya committed:
docs: README 한일 체크

sooyeoniya committed:
refactor: Context를 하나로 통합하는 APIContext 생성

sooyeoniya committed:
refactor: 에러 핸들링 로직 수정

sooyeoniya committed:
refactor: useProducts를 useAPI를 사용해 변경

[세부사항]
useSearchParams를 사용해서 필터링 관리
ProductContext 삭제

sooyeoniya committed:
refactor: useCartItems를 useAPI를 사용해 변경

sooyeoniya committed:
refactor: entities 구조 변경

폴더 내부에 파일이 많지 않으므로 폴더 뎁스 줄이기 위해 하나의 폴더경로로 통일
-> 결론적으로 최대 2개의 뎁스가 감소함

sooyeoniya committed:
chore: MSW API 통신 delay 200 -> 100 으로 줄이기

sooyeoniya committed:
feat: HTTPS 환경에서 MSW 혼합 콘텐츠 오류 해결을 위한 API URL 프로토콜 수정

sooyeoniya committed:
design: 모달 띄워졌을 때도 토스트가 더 위로 올라오도록 z-index 설정

sooyeoniya committed:
refactor: API 반환 타입 명시

sooyeoniya committed:
refactor: ApiSuccess 타입 추가

post, patch, delete와 같은 반환값 없는 요청들을 쉽게 처리하기 위함

sooyeoniya committed:
feat: useApiResponseToasts 훅 구현

응답에 대한 토스트 처리 함수

sooyeoniya committed:
design: QuantitySelector 버튼 opacity 0.4로 변경

sooyeoniya committed:
refactor: get, post, patch, delete에 대한 각각의 api 요청 함수 분리

sooyeoniya committed:
chore: selectedSortOption -> sortOption으로 네이밍 변경

sooyeoniya committed:
chore: React.PropsWithChildren으로 수정

그냥 PropsWithChildren이었는데, import 줄이기 위해 앞에 React 붙임

sooyeoniya committed:
test: useProducts 훅 테스트 수정

sooyeoniya committed:
test: useCartItems 훅 테스트 수정

[문제사항]
refetchCartItems를 spyOn하여 호출되었는지 테스트하는데, 호출이 안되고 있음..

sooyeoniya committed:
test: useCartItems 훅 테스트 오류 수정

sooyeoniya committed:
chore: 헷갈리는 is prefix 제거

JunilHwang commented on src/entities/cartItem/response.ts:
```diff
@@ -1,4 +1,7 @@
+import {
+  BaseProduct,
+  PaginatedResponse,
+} from "../../shared/api/types/response";
```
BaseProduct 라는게 지금 shared 에 있는데,

이름을 토대로 유추해보면 사실 entities에 훨씬 가깝지 않을까요!?

sooyeoniya replied:
해먼드가 위에서 말씀해주신 FSD 폴더 구조 기준을 참고하면서 생각해봤는데요, 해당 타입은 애플리케이션의 도메인 정보를 담고 있는 성격이 강해서, shared에 두기보다는 상품이라는 구체적인 데이터와 직접적으로 연결된 위치에 두는 것이 더 적절하다는 걸 알았습니다!
이번에 전반적으로 폴더 구조 개선 작업을 진행했는데요, 여기서 `src/apis/types/response.ts` 파일로 옮겼고, API 응답과 관련된 타입은 해당 파일에서 일괄적으로 관리하도록 정리해두었습니다!

이번에 전체 코드를 많이 갈아엎는 작업을 진행해서 커밋 내용이 좀 많이 있어서 이해하기 조금 힘들 수 있는 점 양해 부탁드립니다..! 🙇🏻‍♀️ 최대한 관련된 부분을 첨부하도록 해보겠습니다! [응답 타입 파일 위치 변경 커밋](https://github.com/woowacourse/react-shopping-products/pull/138/commits/00f2b5726285f80e5e57093ed5f18a20bd23fd14#diff-11cb5ae251904505040ccd64068ea06927228bef338da93df02d0ea18ded36a8)

JunilHwang commented on src/entities/cartItem/api.ts:
```diff
@@ -0,0 +1,33 @@
+  deleteData,
+  getData,
+  patchData,
+  postData,
```
getData, deleteData 라는 이름은 api 말고 다른 곳에서도 쓰일 수 있다보니..

`apiUtils.getData` 이렇게 namespace 로 묶어주면 어떨까 싶어요..!

sooyeoniya replied:
이번에 API 관련 로직에 대해 너무 복잡하게 관리되고 있다고 판단하여 새롭게 비동기 통신을 관리하는 하나의 [HTTPClient라는 클래스](https://github.com/woowacourse/react-shopping-products/pull/138/commits/c0a3969716d2ba21222e0ee18fe370fd9b96d9c4#diff-e50e37ce231a7522e58249dc77d5a96ad4f4bdca7df229627b963800e2de2b68)를 만들었습니다! 싱글톤처럼 단일 인스턴스만 생성해서 해당 인스턴스만 내보내는 모듈을 하나 만들었습니다!

실제로 사용할 때는 다음 커밋 내용과 같이 적용해주었구요. [CartItemsAPI 코드 수정 ](https://github.com/woowacourse/react-shopping-products/pull/138/commits/c0a3969716d2ba21222e0ee18fe370fd9b96d9c4#diff-39c30c954fefcac582a4f7f78888d6798a17688a46b02c9795db3c92478a7d9a) [ProductsAPI 코드 수정](https://github.com/woowacourse/react-shopping-products/pull/138/commits/c0a3969716d2ba21222e0ee18fe370fd9b96d9c4#diff-008318f3b5eac3b28bd76e3b29b92b3b6c1cb3948066207a67da057b72671269)

이렇게 하나의 클래스에서 공통 get, post, patch, delete 기능을 캡슐화 해서 외부에서 접근할 때 httpClient 하나만 호출하면 되어서, 해먼드가 말한 문제를 해결했습니다!

API URL이랑 호출 시 필요한 개인 인증 토큰 같은 것도, 실질적으로 msw 모킹하는 부분인 handlers를 제외하면 이 모듈 내에서만 쓰이기 때문에 여기서 모두 선언해주었어요!

JunilHwang commented on src/entities/cartItem/api.ts:
```diff
@@ -0,0 +1,33 @@
+import { SHOP_API } from "../../shared/api/config";
+import { CartItems } from "./response";
+
+const URL = SHOP_API.endpoint.cartItems;
```
SHOP_API 라는것도 shared 보단 entities 에 가깝지 않을까요!?

sooyeoniya replied:
이 부분에 있어서도, 비동기 통신 로직을 리팩토링하다보니, 굳이 `SHOP_API`라는 하나의 객체에서 모든 엔드포인트를 일괄적으로 관리할 필요가 없더라구요. 그래서 실제로 사용하는 위치에서 각 엔드포인트를 직접 정의하는 방식으로 변경했습니다! 

아래 커밋을 보면 확인하실 수 있는데요, `cart-items`와 `products`는 각각 CartItemsAPI와 ProductsAPI 내부에서만 사용되기 때문에, 해당 API 모듈 안에서만 관리하도록 했습니다.
다만, msw의 handlers 내부에서는 여러 엔드포인트를 한 번에 다루고 있어서, 엔드포인트들을 모아둔 객체를 내부에서 정의해서 사용해주었습니다!

https://github.com/woowacourse/react-shopping-products/pull/138/commits/26507ff6a5d29fce2ee44912fd2c0c177d8e5319

JunilHwang commented on src/shared/config/base.ts:
```diff
@@ -0,0 +1 @@
+export const BASE_URL = import.meta.env.BASE_URL || "/react-shopping-products";
```
사람마다 shared 라는 폴더에 대한 기준이 다를 수 있는데,

저는 "프로젝트에 종속적이지 않은 코드" 라고 생각해요.
그래서 이런 코드는 shared보단 features 혹은 entities에 더 가까운 코드이지 않을까 싶네요!

sooyeoniya replied:
맞습니다! 해먼드가 알려주신 shared 계층의 책임에 대해 생각해보고나니 이 부분은 shared가 아니라는 것을 이해하게 되었습니다. 
그래서 이번에 폴더 구조를 개편하면서 이 부분을 정확히 어디에 두는 것이 적절할지 고민해봤는데요, 저는 app 계층에 두는 것이 더 적합하다고 판단했습니다! 이 URL은 애플리케이션의 라우팅 경로를 나타내는 URL이기 때문에, features 보다는 앱 전체에서 공통으로 사용되는 설정이라고 판단했습니다! 그래서 app 폴더 하위에서 관리하도록 해주었습니다!

[src/shared/config/base.ts → src/app/config/base.ts](https://github.com/woowacourse/react-shopping-products/pull/138/commits/a90364b84ae29ae161bb4bb3dcc5e69835dac0fa#diff-cd4cbeff9f4275c42a5e21b0a45a6c770e1fc8f5347a745bc31517257d83eaed)

JunilHwang commented on src/shared/api/types/data.ts:
```diff
@@ -0,0 +1,9 @@
+import { CartItems } from "../../../entities/cartItem/response";
+import { Products } from "../../../entities/product/response";
+
+export type ApiDataKey = "products" | "cartItems";
```
products, cartItems 라는 단어도 shared보단 entities 에 어울리는 것 같아요!

sooyeoniya replied:
동의합니다! 다만 이번에 API 관련 데이터 관리 로직을 리팩토링하면서, 해당 부분은 현재는 사용되지 않게 되어 삭제하게 되었습니다 😃

JunilHwang commented on src/shared/hooks/useAPI.ts:
```diff
@@ -0,0 +1,43 @@
+  );
+
+  return { ...result, refetch };
+};
```
https://github.com/user-attachments/assets/11365b28-c73f-4ef2-854b-ebe9ba6332ce

<img width="2036" alt="image" src="https://github.com/user-attachments/assets/b9aec0cb-0f66-450b-97ac-828d9942beaf" />

새로고침할 때 이렇게 많은 API 요청이 이뤄지고 있네요 ㅎㅎ
useAPI의 어떤 부분이 문제일까요!?

sooyeoniya replied:
```tsx
useEffect(() => {
  if (!data[key]) fetchData(key, fetcher);
}, [key, fetcher, data, fetchData]);
```

우선 여기에 `fetcher`가 의존성 배열에 들어가면서, fetcher가 렌더링마다 매번 새로운 함수로 생성될 경우 useEffect가 매번 실행되어서 제가 생각한 의도랑 다르게 API 요청이 많이 보내진 것 같습니다.
fetcher 부분을 useCallback으로 감싸거나, 의존성 배열에서 fetcher를 없애면 될 것 같다는 생각이 듭니다!
어차피 fetcher는 각 key마다 거의 고정된 데이터 패칭 함수이기 때문에 의존성 배열에 fetcher를 없애도 key로 구분이 되어서 괜찮을 것 같습니다.

이번에는 리뷰를 반영하여 의존성 배열 문제를 수정했고, 각 API 요청이 한 번만 발생하도록 안정적으로 동작하게 만들었습니다!
[배포 사이트](https://sooyeoniya.github.io/react-shopping-products/)

JunilHwang commented on src/shared/hooks/useToast.tsx:
```diff
@@ -0,0 +1,12 @@
+import { useContext } from "react";
+import { ToastContext } from "../../app/providers/ToastContext";
```
shared가 app을 참조하고 있네요..!?

sooyeoniya replied:
앞서 FSD 관련해서 말씀해주신 부분을 보고 Toast와 Modal 부분에 대한 파일들의 위치를 모두 수정했는데요!

```
/shared
├── /config
│   └── toast.ts
├── /contexts
│   ├── ModalContext.tsx
│   └── ToastContext.tsx
├── /hooks
│   ├── useModal.tsx
│   └── useToast.tsx
├── /ui
│   ├── Modal
│   │   ├── Modal.styles.ts
│   │   └── Modal.tsx
│   └── Toast
│       ├── Toast.styles.ts
│       └── Toast.tsx
```

위와 같이 shared 계층에서 모두 관련 로직들을 관리할 수 있도록 구성했습니다.

원래는 shared 내부에서도 modal과 toast 관련 파일을 기능 단위로 한 곳에 모아 관리할지(**아래 예시 폴더 구조**) 고민했었는데요,
사실 아래와 같이 구조를 만드는 것이 훨씬 관리하기 편할 것 같다는 생각을 하긴 했습니다.

하지만 현재 상황에서는 다른 폴더들도 모두 관심사별 구조를 따르고 있고,
shared 내부에도 여러 모듈이 있기 때문에, 이를 한꺼번에 관리하려면, 위와 같은 관심사별 구조가 더 효율적이라는 판단이 들었습니다.
따라서 전체 폴더 구조의 일관성을 위해 관심사별로 묶어 관리하는 쪽으로 결정했습니다!

```
/shared
├── /toast
│   ├── ToastContext.tsx
│   ├── useToast.tsx
│   ├── Toast.tsx
│   ├── Toast.styles.ts
│   └── toast.config.ts
├── /modal
│   ├── ModalContext.tsx
│   ├── useModal.tsx
│   ├── Modal.tsx
│   └── Modal.styles.ts
```

JunilHwang commented on src/shared/hooks/useModal.tsx:
```diff
@@ -0,0 +1,12 @@
+import { useContext } from "react";
+import { ModalContext } from "../../app/providers/ModalContext";
```
마찬가지로 shared가 app을 참조하고 있네요..!?

sooyeoniya replied:
위 리뷰 답변에서 말씀드렸다시피 모두 shared 레벨로 옮겨주어, 해당 부분들의 참조 문제를 해결했습니다! :)

JunilHwang commented on src/shared/hooks/useApiResponseToasts.ts:
```diff
@@ -0,0 +1,81 @@
+import { useCallback, useEffect } from "react";
+import { ApiResponse, isApiError, isApiSuccess } from "../api/apiClient";
+import { TOAST_TYPES } from "../config/toast";
+import useToast from "./useToast";
```
useToast가 app을 참조하고,
다시 useToast를 이 훅이 참조하고 있는데요

shared는 어플리케이션의 최하위 계층인데
어플리케이션의 최상위 계층을 참조하고 있다보니 모든 의존성 흐름이 순환구조로 이루어지고 있어요.

이렇게 되면 문제가 뭘까요!?

sooyeoniya replied:
순환 의존성은 계속 이렇게 FSD 계층 구조를 고민하면서 구조를 나눈 의미를 없게 만드는 것 같습니다.
빌드했을 때 이런 순환 의존성으로 인해 모듈 간 로딩 순서가 꼬여 에러가 발생할 수 있기 때문입니다!
그래서 단방향 참조를 유지해야하는 것 같습니다!

sooyeoniya replied:
여기서 궁금한 점이 있습니다 🙋🏻‍♀️

중복되는 비동기 로직을 관리하기 위해 [useFetch](https://github.com/woowacourse/react-shopping-products/pull/138/commits/36d5cf5a3a23ecc9d6ea11ab4eb8c93796479565#diff-7f9b7d69008b45e6ca1c2473af8000c728e746fd45f36059e597b80d4fe551eb)와 [useMutation](https://github.com/woowacourse/react-shopping-products/pull/138/commits/691b5a265e54147f4b8f9a9f0cd0a3c1d7f02b01#diff-253675075913ae91c308e53857adca6963341b158d8f6f8b8074b9d30345491e) 훅을 만들었는데, 이 둘은 데이터 요청과 상태 관리를 담당합니다.
처음에는 여러 위치에서 사용한다는 이유로 shared에 두었지만, 내부에서 useToast를 호출하면서 같은 shared 계층 내에서 훅들이 서로 참조하는 구조가 되어 순환 참조 문제가 발생할 우려가 있더라고요. 물론 useFetch와 useMutation이 useToast를 호출하는 단방향 흐름이긴 하지만요 ㅎㅎ

그래서 src/apis 폴더에 이 두 개의 훅을 두는 게 맞을지 고민이 되는데, 비동기 통신 관련 로직이라 src/apis가 더 적합할 것 같기도 하고, 여러 AI한테 조언을 구해봤는데, 대부분 src/shared에 두는 게 맞다는 의견이었습니다.

따라서, 같은 계층 내에서 훅들이 서로 참조하는 구조는 어떻게 생각하시는지,
이런 구조가 순환 참조 문제를 야기할 가능성이 있어서 좋지 않은 설계인지 궁금합니다!

좋은 방향이나 참고할 만한 설계 방법이 있다면 조언 부탁드립니다!

---

또한, useFetch와 useMutation 내부에서 **useToast를 직접 호출하는 부분**도 조금 걸립니다.

외부에서 토스트 처리를 하는 게 원칙적으로는 맞겠지만, 모든 비동기 호출에서 동일한 방식으로 에러 및 성공 메시지를 보여주기 위해 내부에서 처리하는 편이 중복 로직 없이 깔끔하다고 판단했습니다!

하지만 이렇게 하면 "데이터 패칭/변경"과 "UI 피드백"이 한 곳에 섞여 관심사가 결합되고, 재사용성이 떨어지는 문제도 있는 것 같습니다.

아예 외부에서 onSuccess, onError 를 정의할 때 콜백 함수 내에 showToast를 직접 넣어주는 방식이 더 좋을까요?
아니면 그냥 지금처럼 에러 및 성공 메시지 보여주는 방식이 완전 통일되어 있는 경우에는 지금처럼 중복 로직을 최소화하기 위해 내부에 넣어두는 것이 나을까요?

JunilHwang commented on src/app/providers/ToastContext.tsx:
```diff
@@ -0,0 +1,56 @@
+  };
+
+  return (
+    <ToastContext.Provider value={{ showToast }}>
```
showToast를 useApiResponseToasts에서 사용하고 있는데요,
문제는 useApiResponseToasts에서 showToast가 useMemo, useCallback, useEffect 등에 의존성으로 들어가있다는 점입니다 ㅎㅎ

showToast가 어차피 렌더링 될 때 마다 항상 바뀔텐데 의존성에서 참조하는게 의미가 있는걸까요!?

sooyeoniya replied:
맞는 말씀입니다! 생각해보니 어차피 다 렌더링이 되니까 의미가 없는 부분이었네요!!
우선 `useFetch`와 `useMutation`이라는 훅을 만들어서 구조가 바뀌는 과정에서 `useApiResponseToasts` 이 훅이 사라져버렸지만, 의존성 배열에 대해서도 한 번 더 고려해보면서 구현하는 습관을 가져보겠습니다!! 알려주셔서 감사합니다 😄

JunilHwang commented on src/app/providers/APIContext.tsx:
```diff
@@ -0,0 +1,48 @@
+  );
+
+  return (
+    <APIContext.Provider value={{ data, setData, fetchData }}>
```
컨텍스트를 사용할 때 value의 값에는 memo를 해주는게 중요해요.
이유가 뭘까요!?

sooyeoniya replied:
Context는 value의 `내용`이 아니라 `참조값`을 기준으로 리렌더링 여부를 판단하기 때문에, 실제 내용이 바뀌지 않았더라도 참조값이 매번 바뀌면 하위 컴포넌트들이 불필요하게 리렌더링됩니다.

특히 위에 코드처럼 객체 리터럴을 직접 value로 넘기면, 렌더링될 때마다 새로운 객체가 매번 생성되기 때문에 Context는 매번 value가 변경된 것으로 인식하고, 구독 중인 하위 컴포넌트를 모두 리렌더링하게 되는 문제가 발생하게 됩니다.

그래서 다음과 같이 useMemo로 감싸서 메모이제이션 해주었는데요.

```tsx
  const contextValue = useMemo(
    () => ({
      refetch,
      cartItems: cartItems || [],
      cartItemsCount: cartItems?.length ?? 0,
      totalPriceInCart,
      quantityByProductId,
      decreaseItemQuantity,
      increaseItemQuantity,
      addProductInCart,
      deleteProductInCart,
    }),
    [
      refetch,
      cartItems,
      totalPriceInCart,
      quantityByProductId,
      decreaseItemQuantity,
      increaseItemQuantity,
      addProductInCart,
      deleteProductInCart,
    ]
  );
```
사실 여기서도 한 가지 의문이 들었습니다. 위에 제가 수정한 코드를 보면 useMemo로 value를 감싸고는 있지만, 어차피 의존성 배열에 거의 대부분의 값이 들어가기 때문에, 의존성 중에 하나라도 바뀌면 결국 전체 객체가 새로 만들어지는 건 똑같지 않나 하는 생각이 들었습니다.
그럼에도 useMemo를 해주는 이유는 무엇인지 궁금합니다! 이 수준에서는 아예 컨텍스트를 자주 바뀌는 값과 자주 바뀌지 않는 값으로 분리하거나, 관심사 별로 다시 분리해주는 것이 좋을까요?

JunilHwang replied:
사실 지금은 하는게 큰 의미가 없긴 합니다 ㅎㅎ
왜냐면 많은 컴포넌트가 context를 의존하고 있고,
context의 덩어리가 크기 때문에 작은 변화에도 예민하게 반응하고 있으니까요.

context에 대한 최적화를 하고 싶다면 지금 모습에서 5개 ~ 6개의 context로 분할해야 합니다.
1. 조회와 변이를 분리하여 관리하기.
2. 각각의 값을 context로 분리하여 관리하기.

이렇게 해야 최적화가 가능해요.. ㅎㅎ

제가 memo에 대해 이야기 한 것은 context의 특성에 대해 이해하기를 바라는 마음이었답니다~!

JunilHwang commented on src/entities/product/useProducts.ts:
```diff
@@ -0,0 +1,75 @@
+    [searchParams, setSearchParams]
+  );
+
+  const fetcher = useCallback(
+    () => ProductsAPI.get(category, sortOption),
+    [category, sortOption]
+  );
+
+  const {
+    data: products,
+    error,
+    loading,
+    refetch,
+  } = useAPI("products", fetcher);
```
이 부분의 동작을 이해하는게.. 저는 무척 힘들었어요.

1. 현재 파일에 fetcher 가 정의되어있고
2. useAPI의 callback으로 넘기고
3. useAPI 내부에서 fetchData를 할 때 사용되고
4. 실제로 실행하는건 APIContext 내부에서 실행되고
5. APIContext에서 실행된 결과를 useProducts에서 던져주는 key를 통해 업데이트하고

이런 방식이네요..

이 단계를 어떻게 최소한으로 줄일 수 있을까요?

그리고 Context가 꼭 하나일 필요가 있을까요?
context는 100개여도 상관 없고 1000개여도 상관 없어요 ㅎㅎ
각각의 api마다 한 개의 context를 만들어서 관리할 수 있답니다.

sooyeoniya replied:
저도 사실 이 부분에 있어 해먼드님의 말씀에 공감합니다. 상품 목록 미션의 요구사항 중 하나가 **"Context API 를 활용한다. 단, API 마다 Provider 를 따로 만들지 않고, 하나의 Context 에서 관리할 수 있어야 한다."** 였거든요!
그래서 해당 요구사항을 충족하기 위해 여러 상태를 하나의 Context로 통합해 미션을 진행했었는데, 오히려 그렇게 하다 보니 과도한 추상화로 인해 구조가 더 복잡해지고, 이해하기 어려워진다는 느낌이 강하게 들었습니다.

이번에 리뷰 반영을 하면서 리팩토링을 하는 과정에서도, 이 부분이 너무 이해하기 어려워져서 유지보수하기 어려웠습니다.
그래서, 결국 다시 두 개의 Context로 나누어서 리팩토링을 진행하게 되었어요! 이렇게 하니까 훨씬 더 직관적이고 응집성이 좋다고 느껴졌습니다.

다만, 여전히 고민인 점은 미션 요구사항처럼 Context를 하나로 통합하는 것입니다.
전 무조건적인 추상화보다는, 책임이 명확하게 분리되고 사용성이 좋은 구조가 더 낫다고 생각하는 편이거든요!
그렇다면 왜 이 미션에서는 굳이 하나의 Context로 관리하라는 요구사항이 있었을까? 하는 궁금증이 생겼습니다.

제 나름대로 생각 해봤는데, 사실 해먼드가 슬랙에서도 말씀해주셨다시피, Context는 상태관리용으로 적합하지는 않다고 하셨잖아요!
그때 말씀해주신 useSyncExternalStore 훅을 공부해보면서 Context의 역할에 대해서도 다시 찾아보게 되었습니다.

[이 참고 자료](https://blog.isquaredsoftware.com/2021/01/context-redux-differences/)를 보면, Context는 상태 관리 도구라기보다는 데이터를 컴포넌트에 잘 전달해주기 위한 '파이프' 역할이라는 설명이 있더라구요.

이런 관점에서 생각해보면, 이번 미션에서 "Context를 하나로 관리하라"는 요구사항은 단순히 통합을 강요하려는 것이 아니라, Context를 '하나의 전달 통로'로 바라보며 그 역할을 명확히 이해하게 하려는 목적도 있었던 게 아닐까 하는 생각이 들었습니다.

물론 수업시간에 말씀해주셨던, 기존 Context 사용 방식에서 발생할 수 있는 문제점들(Provider Hell, 불필요한 리렌더링 등)을 경험해보고, 상태 관리 라이브러리가 왜 필요하게 되었는지 직접 체감해보도록 하기 위함도 있지만요!

결론적으로는, Context를 하나로 통합해 데이터를 관리하면서도 어떻게 하면 가독성 좋고, 유지보수하기 쉬운 추상화를 만들 수 있을지에 대한 고민이 있는 것 같습니다. 🥲

JunilHwang replied:
ㅋㅋ 니야가 언급해주신 것 처럼  사실상 context는 상태를 위해 만들어진게 아니라서요..
다만 context의 특성에 대해 잘 이해하기 위해선 이런 과정도 필요하다고 생각합니다.

그리고 제가 요구사항을 잘못알고 있는건지...
컨텍스트는 사용하되 한 개만 사용하라는 부분은 못본 것 같아서요 ㅠㅠ
혹시 수업시간에 언급된 내용일까요?

JunilHwang commented on src/shared/hooks/useApiResponseToasts.ts:
```diff
@@ -0,0 +1,81 @@
+import { TOAST_TYPES } from "../config/toast";
+import useToast from "./useToast";
+
+const useApiResponseToasts = (autoWatchError?: string | null) => {
```
이 훅에 대한 테스트도 있으면 좋지 않을까요!?

sooyeoniya replied:
이 훅이 사라져서 테스트를 못하게 되었습니다. 대신 다른 훅들에 대해서 최대한 테스트 해보겠습니다!

JunilHwang commented on src/shared/hooks/useToast.tsx:
```diff
@@ -0,0 +1,12 @@
+import { ToastContext } from "../../app/providers/ToastContext";
+
+const useToast = () => {
+  const context = useContext(ToastContext);
```
toast에 대한 테스트도 있으면 좋겠네요!

sooyeoniya replied:
Toast 테스트도 구현했습니다! https://github.com/woowacourse/react-shopping-products/pull/138/commits/8a82f2c52dab6554e64026c4ac1e5ce6f111e43c

JunilHwang commented on src/shared/hooks/useModal.tsx:
```diff
@@ -0,0 +1,12 @@
+import { ModalContext } from "../../app/providers/ModalContext";
+
+const useModal = () => {
+  const context = useContext(ModalContext);
```
Modal에 대한 테스트도 있으면 좋을것 같아요..!

sooyeoniya replied:
Modal 관련 테스트 구현했습니다! 확인부탁드립니다 :) https://github.com/woowacourse/react-shopping-products/pull/138/commits/0f67b326441ead4de72bdff3cff107d388924d22

JunilHwang commented on src/widgets/CartModal/CartItem/CartItem.tsx:
```diff
@@ -0,0 +1,36 @@
+  product: { id: productId, name, price, imageUrl, quantity: maxQuantity },
+}: CartItemContent) => {
+  const { increaseItemQuantity, decreaseItemQuantity, deleteProductInCart } =
+    useCartItems();
```
지금 값과 값을 변화시키는 메소드가 다 따로따로 분리되어 있어요.

```tsx
quantity.increment
quantity.decrement
quantity.value
```

이렇게 quantity 라는 관심사에 묶여서 쓰이는 모습을 상상해보면 어떨까요?

product를 다룰 때에도

```tsx
product.deleteInCart(cart.id)
```

이렇게 쓰이는 모습을 상상해보면 좋겠습니다.

sooyeoniya replied:
좋은 방식인 것 같습니다! 해먼드 추천 방식으로 한 번 CartContext와 ProductContext의 value 구조를 변경해봤는데요.

관심사별로 응집된 구조 덕분에, 외부에서 사용할 때도 각 기능에 해당하는 객체 하나만 참조하면 되어 로직이 훨씬 간결해졌습니다!
예를 들어, 아래에 CartContext를 보면 기존에는 2~4개의 값을 따로 호출해야 했었는데, 이제는 하나의 객체로 묶여 있어 코드의 구조가 훨씬 명확하고 보기에도 편해졌습니다.

기능별로 관련된 값과 메서드가 함께 모여 있어 관심사도 더 잘 드러나는 것 같고, 유지보수나 확장 측면에서도 훨씬 가독성이 좋아서 유리하다고 느꼈습니다! 감사합니다 😃

- [CartContext value 구조 변경](http://github.com/woowacourse/react-shopping-products/pull/138/commits/3cc57318d4a9a52406cf7619c63b78034cd7f466)
- [ProductContext value 구조 변경](https://github.com/woowacourse/react-shopping-products/pull/138/commits/d99b75e6d15dd2840f0e16038084ac565fc46930)

JunilHwang commented on src/entities/cartItem/useCartItems.tsx:
```diff
@@ -0,0 +1,158 @@
+  refetchCartItems: () => void;
+}
+
+export const useCartItems = (): UseCartItemsResult => {
```
<img width="1051" alt="image" src="https://github.com/user-attachments/assets/c0345ed3-1009-4506-96ec-2e3a1f2e1f04" />

useCartItems가 여러곳에서 쓰이고 있어요.
각각의 컴포넌트가 렌더링 될 때 마다 API 호출이 발생하겠죠?

한 번의 API 호출만 발생하도록 하고 싶다면 어떤 장치가 필요할까요?

sooyeoniya replied:
[앞선 리뷰](https://github.com/woowacourse/react-shopping-products/pull/138#discussion_r2106675063)와 내용적으로 유사한 리뷰라고 판단했는데요! 

앞서 말씀드렸다시피, useAPI 내부에서 fetcher 가 의존성 배열에 포함되어 있어서 여기서 useCallback으로 감싸더라도 내부적으로 컴포넌트 렌더링마다 fetcher가 새로 만들어지면 매번 effect가 재실행되는 문제가 있었습니다.

그래서 결과적으로 같은 useCartItems 훅을 여러 곳에서 사용하고 있어서 불필요한 중복 API 호출이 발생했던 것 같아요. 이 문제를 해결하려면, 앞선 구조에서는 fetcher를 useEffect의 의존성 배열에서 빼는 게 중복 호출 문제를 해결할 수 있다고 생각했습니다!

---

지금은 새로운 구조로 변경해서 중복 API 호출 문제가 발생하지 않습니다!

우선 useFetch 훅을 통해 데이터를 받아오고, 값을 미리 캐싱 해두기 위해서 [CartProvider](https://github.com/sooyeoniya/react-shopping-products/blob/step2/src/features/cart/contexts/CartContext.tsx)와 [ProductProvider](https://github.com/sooyeoniya/react-shopping-products/blob/step2/src/features/product/contexts/ProductContext.tsx) 내부에서 useState 를 사용해 각 데이터를 저장해두었습니다. (이렇게 안하면, useFetch에서 넘어온 data가 매번 생성되어서, 스크롤이 위로 다시 올라가고 하는 문제가 발생하더라구요..)

그리고 useFetch에서 받아온 data와 success 상태를 가지고 아래와 같이 data가 변경되었고, success 상태가 true로 바뀌었을 때 상태를 저장할 수 있게 수정했습니다.

```tsx
  useEffect(() => {
    if (data && success) {
      setCartItems(data);
    }
  }, [data, success]);
```

여기서 success의 경우에는 data를 받아오면 true가 되고, 다음 API 요청까지는 값이 바뀌지 않습니다. 그러다가 다음 API 호출이 다시 발생했을 때, 요청이 시작되면서 false 상태가 되고 data가 받아와지면 다시 true가 되면서 위의 if 조건문 내부를 실행하게 됩니다.

그리고 초기화 하는 부분도 각 [CartInitializer](https://github.com/sooyeoniya/react-shopping-products/commit/36d5cf5a3a23ecc9d6ea11ab4eb8c93796479565#diff-2c086082cfc931d02f778a95ccea727553e3fa2aeb5aeb5d0b9d5ed28b9547fc)와 [ProductInitializer](https://github.com/sooyeoniya/react-shopping-products/commit/36d5cf5a3a23ecc9d6ea11ab4eb8c93796479565#diff-0c8f652a73d366d0a1004d0f7adc9181b1f3a5ea8577782380561616d6ab74ca) 라는 딱 처음 화면이 렌더링 되는 시점에 데이터를 초기화해주는 컴포넌트를 만들어서, 해당 컴포넌트들을 ProductCatalog(상품 목록) 컴포넌트에 처음에 한 번만 데이터가 받아와질 수 있게 해주었습니다.
이렇게 초기화 컴포넌트를 분리한 이유는 제가 원하는 위치에서 데이터가 받아와질 수 있게 하기 위함이었는데요. 사실 ProductCatalog가 거의 메인 페이지나 다름이 없어서, 그냥 각 Context 내부에서 이 로직을 수행해줘도 무방할 것 같다는 생각이 드네요..

내부 구조는 이런식으로 되어있습니다!

```tsx
const CartInitializer = () => {
  const { refetch } = useCart();

  useEffect(() => {
    if (refetch) refetch();
  }, []);

  return null;
};
```

JunilHwang commented on src/shared/api/apiClient.ts:
중복 로직이 무척 많이 보이네요 ㅎㅎ

이렇게 쓰이는 모습을 상상해보시면 좋겠어요.

```jsx
const cartApiClient = createApiClient({ baseURL: BASE_URL, headers: DEFAULT_HEADERS })

cartApiClient.get(url);
```

중복 로직을 createApiClient 내부에 몰아넣고, 나머지는 get 같은 메소드를 통해 위임하는거죠.

hook을 만드는 방식과 비슷하답니다 ㅎㅎ
hook 내부에서는 컨텍스트를 다 공유하는 것 처럼요!

sooyeoniya replied:
저도 동의합니다! 

일단 저 apiClient 모듈 내부에는 너무 많은 로직들이 한 번에 들어가 있기도 했고, [앞선 리뷰](https://github.com/woowacourse/react-shopping-products/pull/138#discussion_r2106668888)에서도 말씀하셨다시피, getData, postData, deleteData, patchData 가 너무 분산되어서 호출되는 쪽에서 각 메서드를 모두 호출해야하는 문제가 있었던 것 같아요. 

그래서, 이전 구조처럼 **전역 함수로 분산 관리**하는 것이 아니라 **HTTPClient라는 하나의 클래스를 생성**하여 각 메서드를 한 번에 캡슐화하여 관리하도록 수정했습니다! 클래스로 리팩토링하면서 baseUrl과 apiKey를 인스턴스 변수로 관리하도록 구성했고, getHeaders()라는 내부 전용 메서드를 만들어 요청마다 공통 헤더를 생성하도록 했습니다. 이를 통해 각 요청 메서드에서의 중복된 헤더 설정 로직을 깔끔하게 제거할 수 있었습니다!

JunilHwang commented on src/__test__/useCartItems.test.tsx:
```diff
@@ -1,282 +1,518 @@
+const mockHandleError = vi.fn();
+const mockHandleSuccess = vi.fn();
+
+vi.mock("../shared/hooks/useApiResponseToasts", () => ({
```
이 hook을 목킹할 필요가 있는걸까요!? msw가 있어서 나머지는 그냥 써도 무방해보여요.

sooyeoniya replied:
이 훅을 목킹한 이유는, 예를 들어 `expect(mockHandleError).toHaveBeenCalled();`처럼 훅 내부에서 특정 콜백이 실행되었는지를 검증하기 위해서였습니다!

`useApiResponseToasts`는 내부적으로 네트워크 요청을 직접 수행하지는 않고, 토스트 같은 UI 피드백을 트리거하는 역할을 하기 때문에 msw만으로는 각 콜백들의 호출 여부를 검증하기 어려울 거라고 판단했었습니다..

아직 제가 질문의 의도를 제대로 이해를 못한 것 같은데, 혹시 좀 더 구체적으로 설명해주실 수 있을까요..?

JunilHwang replied:
콜백을 실행하는게 목적이라면, useToast를 목킹하는 방식은 어떨까요!?
제일 하위 계층을 목킹하여 사용하는거죠..!

sooyeoniya replied:
오 그 방식이 더 괜찮을 것 같습니다.
실질적으로 토스트가 화면에 잘 띄워지는지를 확인하는 것이 훨씬 더 직접적인 테스트 검증이 될 수 있겠네요!

JunilHwang commented on src/__test__/useCartItems.test.tsx:
```diff
@@ -1,282 +1,518 @@
-    await waitFor(() => {
-      expect(CartItemsAPI.get).toHaveBeenCalled();
-    });
+      await waitFor(() => expect(result.current.loading).toBe(false));
```
watFor을 써야하는 이유가 있을까요?
waitFor을 안 쓰고 표현하는 방법도 고민해주세요..!

그리고 waitFor이 가지고 있는 문제도 같이 고민해보시면 좋겠습니다 ㅎㅎ

sooyeoniya replied:
`waitFor`는 단순히 시간을 기다려주는 역할밖에 하지 않기 때문에, `act`를 사용하거나 `findByText` 같은 것을 사용하는 편이 훨씬 유용할 것 같습니다!

이번에도 리팩토링한대로 테스트를 다시 구현했는데, 이번에는 `waitFor`을 아예 사용하지 않고 구현해봤습니다! 앞서 모달과 토스트에 대한 테스트는 댓글에 남겨두었고, 나머지 테스트는 아래 첨부하겠습니다!

- [useFetch 훅 테스트 구현](https://github.com/woowacourse/react-shopping-products/pull/138/commits/86a3eb5245713b8d653f850adf2b5ea40c23a0af)
- [useMutation 훅 테스트 구현](https://github.com/woowacourse/react-shopping-products/pull/138/commits/53d1fdd087febcdb55ad30a1d982c0af5ba64409)
- [CartContext 테스트 구현](https://github.com/woowacourse/react-shopping-products/pull/138/commits/cdee966cf98fe3581ea596acbb3c70ef8a7422d5)
- [ProductContext 테스트 구현](https://github.com/woowacourse/react-shopping-products/pull/138/commits/3d40a6e896f7e7e360b382937dbf0431503cd3ca)

JunilHwang commented on src/__test__/useCartItems.test.tsx:
```diff
@@ -1,282 +1,518 @@
+      expect(result.current.cartItems?.content[0].product.id).toBe(
+        newProductId
+      );
+      expect(result.current.cartItems?.content[0].quantity).toBe(1);
```
이렇게 테스트가 작성되면 무의미할 확률이 무척 높답니다 ㅠㅠ
이런식으로 모든 값들을 검증하는 방식은 어떨까요?
```tsx
expect(result.current.cartItems).toEqual({
  "content": [
    {
      "id": 101,
        "product": {
        "category": "테스트",
          "id": 1,
          "imageUrl": "test.jpg",
          "name": "테스트용 상품",
          "price": 15000,
          "quantity": 10,
      },
      "quantity": 1,
    },
  ],
})
```

이렇게 표현이 되어야 테스트를 보는 사람들이 동작을 예측하기 더 쉬울 것 같아요!

sooyeoniya replied:
맞습니다. 테스트는 말그대로 '테스트'를 하는 것이기 때문에 과도한 추상화나, 로직을 숨기는 것보다, 테스트 대상을 명시적으로 작성해두는 게 훨씬 유용할 것 같아요!

JunilHwang reviewed (changes requested):
안녕하세요 니야!
상품 목록 2단계 잘 진행해주셨네요 ㅎㅎ 고생하셨습니다.

--

일단 FSD에 대한 리뷰를 모아보자면,
FSD를 정의할 때 각각의 폴더를 니야가 어떤 용도로 생각하고 활용했는지 문서화 하는게 중요해요.
저의 경우 이렇게 정의할 것 같아요.

- shared: 프로젝트와 무관하게 활용할 수 있음
- entities: 도메인 로직이지만 프로젝트와 무관하게 활용할 수 있음. 가령, 로또미션의 lotto 같은 것들
- features: entities + 런타임 종속적인 코드 (로또 미션에서 콘솔 기반으로 구현하는 것과 웹 기반으로 구현할 때를 상상해보면 좋음)
- widgets: features의 조합

그리고 각각의 폴더마다 shared를 만들어서 활용할 수 있답니다.
- features/shared: features에 해당하지만 공통로직인 것들. 가령, 현재 shared/api에 정의된 product, cart 같은 것들

제가 정의한게 정답이라기보단, 제가 프로젝트를 구성할 때 일관성있게 사용할 수 있는 규칙이라고 생각해주세요 ㅎㅎ

중요한건 니야만의 규칙이 필요하고 그 규칙의 근거가 있어야 한답니다.

그리고 꼭 features, entities, shared 같은 것들이 최상위에 있을 필요도 없다고 생각해요.

가령, 이렇게 features의 범위를 조금 더 넓혀서 사용할 수도 있다고 생각합니다. 이러면 관심사 단위로 모듈화 하기가 더 쉽겠죠!?

- utils
- shared/components
- features/products/entities
- features/products/shared
- features/products/ui
- features/products/api
- features/carts/entities
- features/carts/shared
- features/carts/ui
- features/carts/api
- features/shared/api

마지막으로 지금 의존성 규칙이 깨진 모습이 보여지고 있어요.
fsd에서 무엇보다 중요한건 의존성규칙입니다.. ㅎㅎ
단방향 흐름이 유지되록 신경써주세요!

---

api 호출이 지금 불필요하게 많이 발생하고 있습니다. 
이 부분도 꼭 개선해주세요!
자세한건 상세 리뷰를 참고해주세요.

---

질문에 대한 답변을 드리자면

> features를 어떻게 활용해야할지 모르겠습니다.

제가 위에 남긴 상세 리뷰를 참고해주세요!
shared에 있는게 다 features로 가야하지 않나 싶어요.

> Modal / Toast / API 관련 코드 구조가 너무 분산되어 있는 것 같습니다.

마찬가지로 위에 남긴 리뷰를 참고해주세요!
modal, toast 와 관련된 코드는 다 shared에 위치시켜도 좋지 않을까 싶습니다 ㅎㅎ
api에 대한 부분은 더 고민해보시면 좋겠어요!

> ‘is’ 접두사 사용에 대한 고민

is를 사용하지 말자! 가 아니라 is를 일관성있게 사용하자! 라는 의도였답니다 ㅎㅎ
가령 is라는 prefix를 사용한다면..
이런 모습이 될 것 같아요.

```jsx
const isValidUrl = (url: string): boolean =>
  Boolean(url && (url.startsWith("http://") || url.startsWith("https://")));

// isValidUrl 이라는 변수를 어떤식으로 표현해야 좋을까요?
const isValidUrl__var = isValidUrl(url);
```

변수에는 is를 쓰지 않는게 자연스러운 경우가 훨씬 많답니다.

```jsx
const validUrl = isValidUrl(url);
```

> [테스트에서 API 모킹과 실제 호출의 딜레마] 제가 고민하는 부분은, useCartItems 같은 훅에서 다양한 함수들을 사용하는데, API 응답을 모킹하지 않고 제대로 테스트하는 방법을 찾기 어렵다는 점입니다. 제가 잘 활용을 못하고 있는 것 같아서 어떻게 하면 응답 모킹 없이 테스트 오류가 나지 않고 데이터 갱신이 잘 되는 것을 확인할 수 있는지 궁금합니다!

데이터가 갱신되지 않아서 모킹을 사용해주셨다고 이해했어요.
그렇다면 목 함수 자체에서 데이터가 갱신되도록 로직을 작성하면 되지 않을까요!?
아니면 호출 순서를 정의하고, 각각의 호출 순서에 따라 어떤 응답을 보내는지 정의하는 방법도 있답니다!

훅에서 데이터 변화에 따라 어떤 흐름으로 함수들이 호출되는지 검사하는건 유의미하다고 생각해요!

그리고 데이터 스펙에 대한 부분은, api 응답을 아예 픽스쳐로 만들어서 계속 업데이트 하는 방법도 있어요.
이 때 중요한건, 지금은 테스트에서 데이터가 생성되었을 때 "갯수"에 대한 검사를 하고 있는데요, 이러면 높은 확률로 문제가 된답니다 ㅠㅠ
api 응답 결과나 타입이 바뀔 수도 있는데 테스트는 그냥 통과하는 사례여서 어플리케이션에는 문제가 되는데 테스트만 통과하는... 그런 일이 발생할 수 있으니까요.

그래서 테스트를 실행할 때 api 응답에 대한 fixture를 만들어서 관리하는 방법도 고민해보시면 좋아요!


> 아니면 아예 이런 경우에는 API를 직접 호출하는 통합 테스트를 진행하고, 별도로 UI가 잘 업데이트되는지 화면 기반 테스트를 분리해서 하는 것이 더 나을까요?

직접 호출에 대한 부분은 통합테스트라기보단 e2e에 가까울 것 같아요 ㅎㅎ
다만 이 경우에는 test api 가 필요하답니다! 이건 백엔드 개발자와의 협업을 통해서 만들어가야할 것 같아요.

sooyeoniya committed:
refactor: showToast 메모이제이션

sooyeoniya committed:
refactor: mocks와 test 폴더 이름 수정

sooyeoniya committed:
refactor: 전체 폴더 구조 변경

sooyeoniya committed:
refactor: httpClient 구현하여 API 요청 로직 수정

sooyeoniya committed:
refactor: 데이터에 불필요한 값들 제거

sooyeoniya committed:
refactor: useAPI 제거 및 각 도메인 별 컨텍스트 분리

sooyeoniya committed:
fix: setCategory, setSortOption 호출 시 fetchData가 2번 호출되는 문제 해결

setCategory, setSortOption 각각 내부에서 fetchData를 하고 있는데,
fetchData 의존성 자체에도 category, sortOption가 변경되었을 때 다시 호출되므로
둘 중 하나 제거하도록 함
-> setCategory, setSortOption 내부에서 호출하는 fetchData 제거

sooyeoniya committed:
refactor: 불필요한 api key 데이터 제거

sooyeoniya committed:
refactor: useFetch 훅 구현

sooyeoniya committed:
chore: try-catch 로직 추가

sooyeoniya committed:
refactor: useMutation 구현 및 적용

CartContext 내부에 try-catch + showToast 중복 로직이 많아서 useMutation으로 분리함

sooyeoniya committed:
chore: 안쓰는 반환값 제거

sooyeoniya committed:
refactor: products, cart-items End Point 상수 분리

sooyeoniya committed:
chore: 유효하지 않은 테스트 삭제

sooyeoniya committed:
refactor: CartContext value를 도메인 중심의 객체로 묶어 내보내기

sooyeoniya committed:
refactor: ProductContext value를 도메인 중심의 객체로 묶어 내보내기

sooyeoniya committed:
test: Modal 관련 테스트 구현

- Modal 컴포넌트 테스트
- ModalContext와 ModalProvider 테스트 (+useModal 호출)

sooyeoniya committed:
test: Toast 테스트 구현

sooyeoniya committed:
test: useFetch 훅 테스트 구현

sooyeoniya committed:
test: useMutation 훅 테스트 구현

sooyeoniya committed:
test: 테스트 폴더 구조 수정

sooyeoniya committed:
test: CartContext 테스트 구현

sooyeoniya committed:
test: ProductContext 테스트 구현

sooyeoniya commented:
안녕하세요 해먼드! 니야입니다:)
리뷰 반영이 늦어버렸네요.. 죄송합니다 🙇🏻‍♀️

전반적으로, 아래 내용에 대해 초점을 맞춰 리뷰 반영해보았습니다.
- FSD 기반 폴더 구조 개선
- API 통신 로직 개선
- 리팩토링 코드에 대한 테스트 구현

앞서 리뷰 달아주신 부분에 모두 코멘트 달아놓았는데, 확인해주시면 감사하겠습니다!

[🍀 배포 페이지](https://sooyeoniya.github.io/react-shopping-products/)

JunilHwang commented on src/__tests__/CartContext.test.tsx:
```diff
@@ -0,0 +1,209 @@
+  default: () => ({ showToast: mockShowToast }),
+}));
+
+vi.mock("../shared/hooks/useFetch", () => ({
+  default: vi.fn().mockImplementation((fetchFn) => {
+    const fetchData = vi.fn().mockImplementation(async () => {
+      const data = await fetchFn();
+      return { data, success: true };
+    });
+    return { data: initialCartItems, success: true, fetchData };
+  }),
+}));
```
useFetch를 목킹할 필요가 있는걸까요..!?

JunilHwang commented on src/shared/config/fetch.ts:
```diff
@@ -0,0 +1,7 @@
+export const FETCH_TYPES = {
+  START: "START",
+  SUCCESS: "SUCCESS",
+  ERROR: "ERROR",
+} as const;
+
+export type FetchType = (typeof FETCH_TYPES)[keyof typeof FETCH_TYPES];
```
fetch보단 api 혹은 HttpClient 등의 이름으로 표현하면 어떨까 싶어요!

JunilHwang commented on src/shared/contexts/ToastContext.tsx:
```diff
@@ -0,0 +1,55 @@
+  );
+
+  return (
+    <ToastContext.Provider value={{ showToast }}>
+      {children}
+      {!!message && <Toast message={message} type={type} />}
+    </ToastContext.Provider>
```
여러개의 토스트가 필요하다면 어떻게 해야 좋을까요!?

JunilHwang commented on src/shared/hooks/useFetch.ts:
```diff
@@ -0,0 +1,74 @@
+  }
+}
+
+const useFetch = <T, P = void>(
+  fetch: (params: P) => Promise<T>,
+  showError = true
+) => {
+  const { showToast } = useToast();
+  const [state, dispatch] = useReducer(fetchReducer<T>, {
+    data: null,
+    loading: false,
+    error: false,
+    success: false,
+  });
+
+  const fetchData = useCallback(
+    async (params?: P) => {
+      dispatch({ type: FETCH_TYPES.START });
+
+      try {
+        const data = await fetch(params as P);
+        dispatch({ type: FETCH_TYPES.SUCCESS, data });
+      } catch (error: unknown) {
+        const message =
+          error instanceof Error
+            ? error.message
+            : "알 수 없는 오류가 발생했습니다.";
+
+        dispatch({ type: FETCH_TYPES.ERROR, message });
+        if (showError) showToast({ message, type: TOAST_TYPES.ERROR });
+      }
+    },
+    [fetch, showError, showToast]
+  );
+
+  return { ...state, fetchData };
+};
```
`useFecthWithToast` 라는걸 하나 만들어서 관리해주면 어떨까요?

```tsx
const useFetchWithToast = (...props) => {
    const result = useFetch(...props);
    const { showToast } = useToast();

    useEffect(() => {
        if (!error) {
            showToast(error instanceof Error ? error.message : "알 수 없는 오류가 발생했습니다.")
        }
    }, [result.error]);

    return result;
}
```
이런식으로 분리할 수 있을 것 같네요!

JunilHwang commented on src/shared/hooks/useMutation.ts:
```diff
@@ -0,0 +1,46 @@
+  errorMessage?: string;
+}
+
+const useMutation = <T>(
+  mutationFn: (variables: T) => Promise<void>,
+  options?: MutationOptions<T>
+) => {
```
앞선 리뷰와 동일합니다!

JunilHwang replied:
아니면 ErrorBoundary 같은걸 만들어서 관리해줘도 좋을 것 같아요 ㅎㅎ
에러가 발생하면 토스트를 띄워주는 레이어를 만들어놓는거죠.

JunilHwang commented on src/shared/hooks/useMutation.ts:
```diff
@@ -0,0 +1,46 @@
+    [mutationFn, options, showToast]
+  );
+
+  return { mutate };
```
객체로 반환할 필요가 있을까요?

JunilHwang commented on src/shared/ui/Modal/Modal.tsx:
```diff
@@ -0,0 +1,32 @@
+}>;
+
+const Modal = ({ open, onClose, children }: Props) => {
+  const stopPropagation = (e: React.MouseEvent<HTMLDivElement>) => {
+    e.stopPropagation();
+  };
+
```
이런건 유틸함수로 만들어서 관리해주면 어떨까요?
리액트 라이프사이클(state, props 등)과 아예 연관이 없어서요!

JunilHwang commented on src/shared/ui/Modal/Modal.tsx:
```diff
@@ -0,0 +1,32 @@
+  onClose: () => void;
+}>;
+
+const Modal = ({ open, onClose, children }: Props) => {
```
onClose 라는 이름은 close가 발생할 때 이를 밖에 알려주는 이벤트를 의미하는데요
즉, close 기능이 modal에 있어야 합니다.
그런데 지금은 close 기능 자체를 onClose가 수행하고 있어요.

그렇다면 다른 이름이 필요한게 아닐까요!?

JunilHwang commented on src/shared/contexts/ModalContext.tsx:
```diff
@@ -0,0 +1,27 @@
+  return (
+    <ModalContext.Provider value={{ openModal, closeModal }}>
+      {children}
+      <Modal open={modalContent !== null} onClose={closeModal}>
```
modal에게 props를 넘겨줄게 아니라 내부에서 context로 값을 받아와서 close를 수행하고
그 다음에 바깥쪽에 알려야 하는게 아닐까요?
```tsx
const Modal = ({ onClose }) => {
    const { close } = useModal();

    // close를 실행하고, onClose가 있으면 밖에 알림
    const handleClose = () => {
        close();
        onClose?.();
    }
}
```

JunilHwang commented on src/shared/utils/image.ts:
```diff
@@ -0,0 +1,7 @@
+const DEFAULT_IMAGE_URL = "./default-image.webp";
+
+const isValidUrl = (url: string): boolean =>
+  Boolean(url && (url.startsWith("http://") || url.startsWith("https://")));
+
+export const getImageUrl = (url: string): string =>
+  isValidUrl(url) ? url : DEFAULT_IMAGE_URL;
```
여기서 return type 은 정의하지 않아도 된답니다.
이유가 뭘까요!?

JunilHwang commented on src/apis/types/response.ts:
```diff
@@ -0,0 +1,14 @@
+export interface Product {
+  id: number;
+  name: string;
+  price: number;
+  imageUrl: string;
+  category: string;
+  quantity: number;
+}
+
+export interface CartItem {
+  id: number;
+  quantity: number;
+  product: Product;
+}
```
api가 아니라 feature에 혹은 domain에 정의되어야 하지 않을까요!?
가령, localStorage 같은곳에서 이 타입에 해당하는 값들이 쓰여야 한다면 어떻게 해야 좋을까요?

JunilHwang commented on src/apis/httpClient.ts:
```diff
@@ -0,0 +1,69 @@
+    this.apiKey = apiKey || "";
+  }
+
+  private getHeaders() {
+    return {
+      "Content-Type": "application/json",
+      ...(this.apiKey && { Authorization: `Basic ${this.apiKey}` }),
+    };
+  }
+
```
```suggestion
  private get headers() {
    return {
      "Content-Type": "application/json",
      ...(this.apiKey && { Authorization: `Basic ${this.apiKey}` }),
    };
  }
```
getter로 정의해주면 더 편할 것 같아요!

JunilHwang commented on src/apis/httpClient.ts:
```diff
@@ -0,0 +1,69 @@
+    return response;
+  }
+
+  setApiKey(apiKey: string) {
+    this.apiKey = apiKey;
+  }
```
안쓰이는 메소드네요..!

JunilHwang commented on src/apis/httpClient.ts:
```diff
@@ -0,0 +1,69 @@
+};
+
+export const API_BASE_URL = getApiBaseUrl();
+export const API_TOKEN = import.meta.env.VITE_TOKEN;
```
외부에서는 안쓰이고 있네요!
export 는 안 해줘도 무방해보여요~

JunilHwang commented on src/apis/httpClient.ts:
```diff
@@ -0,0 +1,69 @@
+const getApiBaseUrl = () => {
+  if (import.meta.env.VITE_USE_MOCK === "true") {
+    return "https://techcourse-lv2-alb-974870821.ap-northeast-2.elb.amazonaws.com/";
```
목킹할땐 그냥 "/" 하나만 반환해도 똑같지 않을까요!?

JunilHwang commented on src/widgets/ShopHeader/ShopHeader.tsx:
```diff
@@ -0,0 +1,25 @@
+  const { cart } = useCart();
+  const { openModal } = useModal();
+
+  const handleOpenCart = () => openModal(<CartModal />);
```
handleOpenCart => openCartModal
이벤트를 다루고 있지 않기 때문에 이렇게 표현해도 좋을 것 같아요..!

JunilHwang commented on src/widgets/CartModal/CartModal.tsx:
```diff
@@ -0,0 +1,29 @@
+          {cart.totalPrice.toLocaleString()}원
+        </S.TotalPriceValue>
+      </S.TotalPriceContainer>
+      <S.CloseButton onClick={closeModal}>닫기</S.CloseButton>
```
```tsx
<CloseTrigger>
    <S.CloseButton>닫기</S.CloseButton>
</CloseTrigger>
```
이런 모습을 상상해보시면 어떨까요!?
이런걸 headless ui 라고 이야기 한답니다 ㅎㅎ

참고: https://www.radix-ui.com/themes/docs/components/dialog

JunilHwang commented on src/widgets/CartModal/CartItem/CartItem.tsx:
```diff
@@ -0,0 +1,35 @@
+  quantity: currentQuantity,
+  product: { id: productId, name, price, imageUrl, quantity: maxQuantity },
+}: CartItemType) => {
+  const { product } = useCart();
```
CartItem 밖에서도 useCart로 cart.Items 를 받아오고 있어요.
그렇다면, 어차피 useCart를 사용하는김에 지금도 props로 값들을 받아올게 아니라, 이런 모습으로 쓰일 수 있지 않을까요?

```tsx
const { cart, product } = useCart();
const { quantity, productId, name, price, imageUrl, maxQuantity } = getProduct(cart, id);
```

요로코롬...

JunilHwang commented on src/features/cart/apis/CartItemsAPI.ts:
```diff
@@ -0,0 +1,50 @@
+const ENDPOINT = "cart-items";
+
+export const CartItemsAPI = {
+  get: async (): Promise<CartItem[]> => {
+    const params = new URLSearchParams({
+      page: "0",
+      size: "50",
+      sort: "asc",
+    });
+
+    const response = await httpClient.get(`${ENDPOINT}?${params.toString()}`);
+    if (!response.ok) throw new Error(ERROR_MESSAGE.GET);
+
+    return await response.json();
+  },
```
```suggestion
  get: async () => {
    const params = new URLSearchParams({
      page: "0",
      size: "50",
      sort: "asc",
    });

    const response = await httpClient.get<CartItem[]>(`${ENDPOINT}?${params.toString()}`);
    if (!response.ok) throw new Error(ERROR_MESSAGE.GET);

    return await response.json();
  },
```
이렇게 generic을 통해 타입을 지정할 수 있도록 httpClient를 정의해주면 어떨까요?

JunilHwang commented on src/features/cart/contexts/CartContext.tsx:
```diff
@@ -0,0 +1,191 @@
+
+export const CartProvider = ({ children }: PropsWithChildren) => {
+  const [cartItems, setCartItems] = useState<CartItem[]>([]);
+  const { data, success, fetchData } = useFetch<CartItem[]>(() =>
+    CartItemsAPI.get()
+  );
+
+  useEffect(() => {
+    if (data && success) {
+      setCartItems(data);
+    }
+  }, [data, success]);
+
+  const refetch = useCallback(() => fetchData(), [fetchData]);
```
fetchData는 매번 바뀐답니다.
그래서 useCallback으로 감싸줄 필요가 사실... 없어요.
이유가 뭘까요?
그리고 refetch가 이렇게 정의되어도 똑같을 것 같네요!

```tsx
const refetch = fetchData
```

JunilHwang commented on src/shared/hooks/useFetch.ts:
```diff
@@ -0,0 +1,74 @@
+  }
+}
+
+const useFetch = <T, P = void>(
+  fetch: (params: P) => Promise<T>,
+  showError = true
+) => {
+  const { showToast } = useToast();
+  const [state, dispatch] = useReducer(fetchReducer<T>, {
+    data: null,
+    loading: false,
+    error: false,
+    success: false,
+  });
+
+  const fetchData = useCallback(
+    async (params?: P) => {
+      dispatch({ type: FETCH_TYPES.START });
+
+      try {
+        const data = await fetch(params as P);
+        dispatch({ type: FETCH_TYPES.SUCCESS, data });
+      } catch (error: unknown) {
+        const message =
+          error instanceof Error
+            ? error.message
+            : "알 수 없는 오류가 발생했습니다.";
+
+        dispatch({ type: FETCH_TYPES.ERROR, message });
+        if (showError) showToast({ message, type: TOAST_TYPES.ERROR });
+      }
+    },
+    [fetch, showError, showToast]
+  );
```
방금전에 CartContext에서 fetchData를 useCallback의 deps로 정의했는데요,
여기서 훅에서 받아오는 fetch 함수를 잘 살펴봐야합니다.
fetch 함수가 변경되면 fetchData를 다시 만들고 있어요.

JunilHwang commented on src/features/cart/contexts/CartContext.tsx:
```diff
@@ -0,0 +1,191 @@
+
+export const CartProvider = ({ children }: PropsWithChildren) => {
+  const [cartItems, setCartItems] = useState<CartItem[]>([]);
+  const { data, success, fetchData } = useFetch<CartItem[]>(() =>
+    CartItemsAPI.get()
+  );
```
그런데 이 구간을 살펴보면, useFetch의 인자로 넘겨주는 함수가 익명함수입니다.
즉, 훅이 실행될 때 마다 재선언 되는거죠 ㅎㅎ

이 부분 때문에, 이와 관련된 모든 메모이제이션이 다 깨지게 됩니다.

이를 어떻게 해결할 수 있을까요?
제일 간단한 방법(useCallback이나 useMemo를 사용하지 않고 해결할 수 있는 방법)을 생각해보세요!

JunilHwang commented on src/features/cart/contexts/CartContext.tsx:
```diff
@@ -0,0 +1,191 @@
+    onSuccess: () => fetchData(),
+  });
+
+  const { mutate: addToCart } = useMutation<number>(
+    (productId) => CartItemsAPI.post(productId),
+    {
+      successMessage: "상품이 장바구니에 추가되었습니다.",
+      onSuccess: () => fetchData(),
+    }
+  );
```
useMutation에 넘겨주는 인자들도 항상 다시 만들어집니다.
그런데 useMutation 내부에서는 이 인자들을 토대로 메모이제이션을 하고 있네요 ㅎㅎ

JunilHwang commented on src/features/cart/contexts/CartContext.tsx:
```diff
@@ -0,0 +1,191 @@
+    },
+    [addToCart]
+  );
+
+  const deleteProductInCart = useCallback(
+    async (cartId: number) => {
+      await deleteCartItem(cartId);
+    },
+    [deleteCartItem]
+  );
+
+  const contextValue = useMemo(
+    () => ({
+      refetch,
+      cart: {
+        items: cartItems,
+        count: cartItems?.length ?? 0,
+        totalPrice: totalPriceInCart,
+      },
+      product: {
+        add: addProductInCart,
+        delete: deleteProductInCart,
+        quantity: {
+          get: quantityByProductId,
+          increase: increaseItemQuantity,
+          decrease: decreaseItemQuantity,
+        },
+      },
+    }),
+    [
+      refetch,
+      cartItems,
+      totalPriceInCart,
+      quantityByProductId,
+      decreaseItemQuantity,
+      increaseItemQuantity,
+      addProductInCart,
+      deleteProductInCart,
+    ]
+  );
```
앞선 내용들을 토대로 유추해보자면,
여기에 정의된 모든 메모이제이션은 결국 다 소용이 없답니다..

JunilHwang commented on src/features/cart/contexts/CartInitializer.ts:
```diff
@@ -0,0 +1,14 @@
+  const { refetch } = useCart();
+
+  useEffect(() => {
+    if (refetch) refetch();
```
```suggestion
    refetch?.();
```

이렇게 표현할 수 있답니다!

JunilHwang commented on src/features/product/apis/ProductsAPI.ts:
```diff
@@ -0,0 +1,33 @@
+    category: CategoryOptionsKey,
+    sortOption: SortOptionsKey
+  ): Promise<Product[]> => {
+    const params: Record<string, string> = { page: "0", size: "20" };
```
```suggestion
    const params = { page: "0", size: "20" };
```
일러 땐 타입을 정의하지 않아도 된답니다 ㅎㅎ 오히려 정의하지 않았을 때 유추되는 타입이 더 정확해요.

JunilHwang commented on src/features/product/apis/ProductsAPI.ts:
```diff
@@ -0,0 +1,33 @@
+    category: CategoryOptionsKey,
+    sortOption: SortOptionsKey
+  ): Promise<Product[]> => {
+    const params: Record<string, string> = { page: "0", size: "20" };
+
+    if (category !== "전체") params.category = category;
+    if (sortOption && sortOptionsMap[sortOption]) {
+      params.sort = sortOptionsMap[sortOption];
+    }
```
```suggestion
    const params = {
        page: "0",
        size: "20"
        ...(category !== "전체" && { category }),
        ...(sortOption && sortOptionsMap[sortOption] && { sort: sortOptionsMap[sortOption] }),
    };
```
이렇게 표현할 수 있답니다!

JunilHwang commented on src/features/product/apis/ProductsAPI.ts:
```diff
@@ -0,0 +1,33 @@
+      params.sort = sortOptionsMap[sortOption];
+    }
+
+    const response = await httpClient.get(
+      `${ENDPOINT}?${new URLSearchParams(params).toString()}`
+    );
```
```suggestion
    const response = await httpClient.get(ENDPOINT, { query: params });
```
이렇게 사용할 수 있도록 httpClient 내부를 추상화해주면 어떨까요?

JunilHwang commented on src/features/product/contexts/ProductContext.tsx:
```diff
@@ -0,0 +1,121 @@
+  const [products, setProducts] = useState<Product[]>([]);
+  const [searchParams, setSearchParams] = useSearchParams();
+
+  const fetchProducts = useCallback(
+    (params: ProductParams) =>
+      ProductsAPI.get(params.category, params.sortOption),
+    []
+  );
```
이건 리액트와 아예 연관이 없는 함수인데요,
컴포넌트 밖에 정의해주면 어떨까요?

JunilHwang commented on src/features/product/contexts/ProductContext.tsx:
```diff
@@ -0,0 +1,121 @@
+    []
+  );
+
+  const { data, loading, success, fetchData } = useFetch<
+    Product[],
+    ProductParams
+  >(fetchProducts);
```
함수를 통해 타입이 자동으로 유추되도록 하려면 어떻게 해야 좋을까요?

```suggestion
  const { data, loading, success, fetchData } = useFetch(fetchProducts);
```

이렇게 쓰고 싶은거죠 ㅎㅎ

JunilHwang commented on src/features/product/contexts/ProductContext.tsx:
```diff
@@ -0,0 +1,121 @@
+      loading,
+      products,
+
+      category: {
+        value: category,
+        set: setCategory,
+      },
+
+      sort: {
+        value: sortOption,
+        set: setSortOption,
+      },
```
네임스페이스로 잘 묶어주셨네요 ㅎㅎ 보기 좋습니다.

JunilHwang reviewed (approved):
안녕하세요 니야!
피드백 반영 꼼꼼하게 잘 해주셨네요 ㅎㅎ 고생하셨습니다!
질문주신 내용에도 답변 달아놓았으니 한 번 확인 부탁드립니다.

---

지금 메모이제이션에 대한 부분이 아직 익숙하지 않은게 느껴져요!
꼼꼼하게 적용해주셨으나 사실상 효과가 없게 되었네요 ㅠㅠ
아마 익숙해지려면 시간이 조금 필요하지 않나 싶습니다.

근데 어차피 context의 덩어리가 커서 메모이제이션을 완벽하게 적용했다고 해도 큰 효과가 없을 확률이 굉장히 높답니다 ㅎㅎ

그래서 이럴 때는 상태관리 라이브러리를 사용하는게 효과적입니다.
이런 과정을 거치면서 context의 특성에 대해 알아가고, 어떤 상황에 쓰면 좋은지 실패를 통해 알아가는거라고 생각해요 ㅋㅋ

이 외에도 자잘한 피드백 남겼으니 다음 단계 진행하실 때 참고 부탁드려요! (사실상 막바지네요..)

...

부족한 리뷰어였으나 조금이나마 니야에게 도움이 되었길 바랍니다 ㅎㅎ
이번 미션은 여기서 마무리할게요! 고생하셨습니다!
